[MRG] Fix numpy.int overflow in make_classification

VarIr opened the pull request:
The sample generator `make_classification()` raises misleading errors on certain valid inputs.

#### What does this implement/fix? Explain your changes.

The sample generator `make_classification()` checks its parameters,
e.g. `2 ** n_informative < n_classes * n_clusters_per_class`.

If `n_informative` is given as numpy.int with a value of 64 or larger,
`2 ** n_informative` evaluates to 0, and the check fails with a misleading error message.

Casting to Python int() avoids this issue.

#### Reproduce the error
```python
>>> import numpy as np
>>> from sklearn.datasets import make_classification 
>>> N_INFORMATIVE = np.arange(31, 65)
>>> for n_informative in N_INFORMATIVE:
>>>      print(f'n_informative = {n_informative}, 2 ** n_informative = {2 ** n_informative}')
>>>      make_classification(n_features=100, 
>>>          n_informative=n_informative, n_classes=2, n_clusters_per_class=1)
```

#### Any other comments?
`2. ** n_informative` would also do the trick.

VarIr committed:
FIX numpy.int overflow in make_classification

Sample generator `make_classification()` checks that `2 ** n_informative < n_classes * n_clusters_per_class`.
If `n_informative` is given as numpy.int with a value of 64 or larger, `2 ** n_informative` evaluates to 0, and the check fails with a misleading error message.
Casting to Python int() avoids this issue.

jnothman commented on sklearn/datasets/samples_generator.py:
```diff
@@ -161,7 +161,7 @@ def make_classification(n_samples=100, n_features=20, n_informative=2,
         raise ValueError("Number of informative, redundant and repeated "
                          "features must sum to less than the number of total"
                          " features")
-    if 2 ** n_informative < n_classes * n_clusters_per_class:
```
Or we could make 2 a float, or use log2 on the right hand side, or cast n_informative to a python int.

VarIr replied:
Float 2. ** n_informative overflows at n_informative = 1024.

But we might go better with log2, since this should avoid any overflows (for sensible n_classes, n_clusters), and might be a tad faster than operations on large Python ints.

lesteve replied:
+1 for using log2

jnothman reviewed (commented):
This should have a non-regression test

VarIr committed:
Using log2

VarIr committed:
TST make_classification() with numpy.int

Parameter n_informative as numpy.int

jnothman reviewed (commented):
We should have a non-regression test, and perhaps a comment explaining why we have not used the simpler expression

VarIr commented:
The tests fail, because the hypercube samples generator does not create centered data down to the precision required by the tests. This happens for both numpy and Pythons integers. The generator behaves differently when n_informative > 30. These cases are currently covered by `test_make_classification()`, but not by `test_make_classification_informative_features()`. 

Should we allow less precision in the tests for centered hypercube data? Otherwise, we need to take a closer look on the hypercube generator.

@jnothman Please specify what test you have in mind.

lesteve commented:
A non-regression test is a test that fails with master and that does not fail in your PR.

lesteve commented:
Quickly looking at it it looks like the test you added is fine. I see a shape mismatch at the moment, so is it really a precision problem:
```
E                   AssertionError: 
E                   Arrays are not almost equal to 0 decimals
E                   Clusters are not centered on hypercube vertices
E                   (shapes (64,), (1,) mismatch)
E                    x: array([ 1000000.68194634,  1000002.57310294,   999999.46742862,
E                           1000000.84695066,   999999.7296249 ,  1000000.21288462,
E                           1000000.79945417,  1000000.3499359 ,  1000000.36114583,...
E                    y: array([ 64000000.])
```

If you can make it pass with lowering the precision this may be an option.

VarIr committed:
Comment on log2

VarIr committed:
Lower hypercube center precision

VarIr commented:
The shape mismatch comes from `np.array(64)` instead of `np.int(64)`... After fixing that, it's still a precision issue.

Also, I just found that for large n_instances and small n_classes (e.g. 1080, 2) the cluster assertion fails. Increasing class_sep seems to solve this, but I will test a bit more.

VarIr committed:
fix pyflakes issue

VarIr commented:
From my point of view, this is ready to go.

jnothman reviewed (approved):
Thanks for the reminder

VarIr commented on sklearn/datasets/tests/test_samples_generator.py:
```diff
@@ -128,9 +129,10 @@ def test_make_classification_informative_features():
-                                              [class_sep] * n_informative,
-                                              decimal=0,
+                    assert_array_almost_equal(np.abs(centroid) / class_sep,
+                                              np.array([class_sep]
```
`np.array([n_informative])` is an array of shape `(1,)`, but we need an array of shape `(n_informative,)`.
Anyway, the mixture of Python list operations and numpy is confusing,
and I guess it should be `np.ones(n_informative)` really..

jnothman replied:
Yes, that's it

VarIr committed:
Simplification and consistency

Replaced confusing Python list operations with concise numpy statement

Consistent assertions in both hypercube branches

VarIr commented:
For matters of consistency and removal of confusing Python list operations, I also changed the `assert_array_almost_equal` in `not hypercube` branch.
In both branches, precision is thus reduced from 6 to 5 decimals.

VarIr commented:
A friendly reminder of this little PR. Any comments, @lesteve?

rth commented:
Thanks @VarIr !

qinhanmin2014 commented:
@VarIr Please add an entry to the change log at `doc/whats_new/v*.rst`. Like the other entries there, please reference this pull request with `:issue:` and credit yourself (and other contributors if applicable) with `:user:` (We have entries for other similar issues, e.g., #10414, #10844).
I guess we can hurry this one into 0.20.1.

rth commented:
@qinhanmin2014 this got merged already, sorry. I was also considering asking for a what's new entry, in the end incorrectly decided against it (and merged) -- for some reason I got confused about the end effect of this (though it only addressed "misleading error messages").

You are right, it would deserve a what's new, since this prevents producing errors for input that is valid.

@VarIr If you can add a what's new in a comment, we will integrate it.

VarIr commented:
Thank you all!

@rth Here's my what's new:
Fixed a bug in :func:`datasets.make_classification` to avoid integer overflow on certain valid inputs of numpy dtypes. Logarithmized an inequation to avoid exponentiation with a potentially very high number.
:issue:`10811` by :user:`Roman Feldbauer <VarIr>`

qinhanmin2014 commented:
@VarIr Please submit a PR, thanks.

rth commented:
Added the what's new, somewhat adapted, in https://github.com/scikit-learn/scikit-learn/commit/da858158bde28f2ceb9982d57cf2ef9636039c3f.  Thanks all!
